### PR TITLE
Fix throttled status updates by use `QSignalThrottler` in `_QtMainWindow` constructor

### LINKED ISSUE #4982: Bug: Segfaults appearing with "QObject::setParent: Cannot set parent, new parent is in a different thread"
## 🐛 Bug

In the last couple days, I've seen this segfault a couple times on main.  I'm not quite sure what's causing it yet, but it tends to happen with mouse movements, perhaps on the canvas, maybe on sliders, not sure...  but it crashes napari out to the console

```
WARNING: QObject::setParent: Cannot set parent, new parent is in a different thread
10:19:54 WARNING QObject::setParent: Cannot set parent, new parent is in a different thread
WARNING: QObject::setParent: Cannot set parent, new parent is in a different thread
10:19:54 WARNING QObject::setParent: Cannot set parent, new parent is in a different thread
WARNING: QObject::setParent: Cannot set parent, new parent is in a different thread
10:19:54 WARNING QObject::setParent: Cannot set parent, new parent is in a different thread
[1]    19897 segmentation fault  napari
```

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This is an alternative implementation to #4983, that reconnects `viewer._update_status_bar_from_cursor` in `_QtMainWindow` constructor and keeps headless mode working. 

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

closes #4982

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/_qt/qt_main_window.py
Comment (andy-sweet) [inline napari/_qt/qt_main_window.py]: Interesting idea!

I guess this means if I do `viewer.cursor.position = ...` in the console/script in non-headless napari, then we won't get the side-effects anymore. It's a bit weird to have this inconsistency, but otherwise I don't mind this.

Comment (andy-sweet) [reply napari/_qt/qt_main_window.py]: Oh wait, I'm wrong. The connection below will still handle that.

Comment (Czaki) [reply napari/_qt/qt_main_window.py]: see the line below when I again connect the same function, but in throttled version. 

Or what do you mean by side-effect?

Comment (andy-sweet) [reply napari/_qt/qt_main_window.py]: Though, I think there should be an explanation in the code that tells the reader why we do this reconnection (because we want to limit it in the GUI).

**Review by andy-sweet — APPROVED** [review]
Looks good to me - just one line comment and [the failing test](https://github.com/napari/napari/runs/8024496473?check_suite_focus=true#step:8:60) to resolve.

File: napari/_tests/test_multiple_viewers.py
Comment (Czaki) [inline napari/_tests/test_multiple_viewers.py]: @tlambert03 Why is this test using `Viewer` directly instead of `make_napari_viewer`? It was added in #3692

File: napari/_qt/qt_main_window.py
Comment (andy-sweet) [inline napari/_qt/qt_main_window.py]:
```suggestion
        # In the GUI we expect lots of changes to the cursor position, so
        # replace the direct connection with a throttled one.
```

The explanation is the key thing here - the code mostly describes itself already.

File: napari/_tests/test_multiple_viewers.py
Comment (andy-sweet) [inline napari/_tests/test_multiple_viewers.py:7]: Did this test start causing problems after the changes here, or is it independently flaky?

File: napari/_tests/test_multiple_viewers.py
Comment (andy-sweet) [inline napari/_tests/test_multiple_viewers.py:23]: If there's a specific explanation about why we need the arbitrary wait at the end (e.g. the 50ms comes from the throttled timeout duration), then add it. If not, then I'll just close my eyes and ignore this for now 🙈

File: napari/_qt/qt_main_window.py
Comment (andy-sweet) [inline napari/_qt/qt_main_window.py:297]: I guess `self.status_throttler.cancel()` isn't enough here right? Would be nice to have a better alternative than this, but not required.

**Review by andy-sweet — APPROVED** [review]
Still looks fine to me, but should get @tlambert03's review before merging, as this is my first time thinking about throttling Qt.

File: napari/_tests/test_multiple_viewers.py
Comment (Czaki) [reply napari/_tests/test_multiple_viewers.py:7]: This is the only test in which we use `Viewer` instead of `make_napari_viewer` so it does not go through this check:
https://github.com/napari/napari/blob/f6bdd6235da23d4d23239653b6e4cf02e7f6d4ac/napari/utils/_testsupport.py#L233-L236 

 So there is only one place where viewers may leak.

File: napari/_qt/qt_main_window.py
Comment (Czaki) [reply napari/_qt/qt_main_window.py:297]: https://github.com/napari/superqt/blob/e7e8dfc44c4c984126e52385015eed4942519f87/src/superqt/utils/_throttler.py?rgh-link-date=2022-08-26T23%3A08%3A47Z#L138-L140

no, because cancel does not stop the timer. It only switches the binary flag.

File: napari/_tests/test_multiple_viewers.py
Comment (Czaki) [reply napari/_tests/test_multiple_viewers.py:23]: I try to follow this
https://github.com/napari/napari/blob/f6bdd6235da23d4d23239653b6e4cf02e7f6d4ac/napari/utils/_testsupport.py#L173

but decide to stop this debug and wait on @tlambert03 response to if he knows why we use directly `Viewer` here.

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py:297]: Just curious, why does it matter if the timer keeps running if it does nothing when it times out? (Which is what cancel() does)

File: napari/_tests/test_multiple_viewers.py
Comment (tlambert03) [reply napari/_tests/test_multiple_viewers.py]: Can't remember,  but looking at it now, my guess is to make it as similar to whatever bug we were fixing in the PR that added this test.  That is, to avoid any addition differences that make_napari_viewer might have caused. 

is it important to change it here?  Or just curious

Comment (Czaki) [reply napari/_tests/test_multiple_viewers.py]: without skip this test, the whole test set fails because of the leaked widget.

File: napari/_qt/qt_main_window.py
Comment (Czaki) [reply napari/_qt/qt_main_window.py:297]: Qt may keep objects alive if events need to proceed. So the started timer may keep the parent object alive until the timer is finished. But this is only my guess, I have no valid proof.

**Review by tlambert03 — APPROVED** [review]
fine with me too.  thanks @Czaki.  
Sorry about the merge conflicts now.  feel free to revert and/or overwrite where needed

**Comment by alisterburt** [comment]
Great work @Czaki and thanks for the reviews @andy-sweet and @tlambert03 - this would be great to get in for 0.4.17. @Czaki if you find time to fix the merge conflicts that would be great, we can then merge here

**Comment by alisterburt** [comment]
This looks great and all tests are now passing @Czaki - I will merge in a few hours unless I hear otherwise!

**Pull request merged**
